### DISCUSSION TITLE
Nodal boundary conditions via imported CSV

### DISCUSSION THREAD

**Discussion by Traiwit** [idaholab/moose#21209] [Q&A General] [answered]
Hi guys,

I've been looking at Nodal boundary conditions, unfortunately I couldn't find it any where on MOOSE source page.

So the idea would be this, I have a csv table of nodes and their DirichletBC value, not sure how should I implement this into MOOSE.
![image](https://user-images.githubusercontent.com/79562439/171773496-8bb49271-4d39-4507-9f43-5ff9a1a1f589.png)

As far as I can see, most of the MOOSE BCs require 'boundary', but I don't want this, I don't want to define any surface.

**************************************************************************************************************************************
1
My initial idea would be this: import CSV file via `PropertyReadFile` but the problem with this method would be that I would have to prepare the CSV file with all the nodes in the domain (rather than just the nodes at boundary/surface) + I don't want to impose any value outside the boundary/surface even with 0 value

then use `PiecewiseConstantFromCSV` to store the value from `PropertyReadFile` > then use `FunctionDirichletBC` to impose the BC values, but again `FunctionDirichletBC` requires boundary name/ID

2
Another more silly approach would be that I generate 100 of thousand boundaries using `SideSetsFromPointsGenerator` or pre-generate it elsewhere, and assign DirichletBC to each one of them (can be done via an action too).

3
Another method would be to create my own Nodal boundary conditions.

If you have any idea or have implemented this, please let me know.

Thank you!

Kind regards,
Traiwit

**Comment by Traiwit** [comment]
I see that there are NodalBC.C and NodalBCBase.C in the source files, which might be useful, I will investigate further

**Comment by GiudGiud** [comment, marked as answer]
Hello

I think 1 would work out of the box.
Note that you can specify bogus values for all the nodes that you dont care about in the CSV file. The only values that will be used by a FunctionDirichletBC would be the nodes on the boundary. It doesnt query anywhere else, the values would be ignored. 

2 will not help you with your memory issues

3 would be the cleaner one, calling the propertyReadFile directly from the BC object. As you prefer though. 

Guillaume

**Reply by Traiwit** [reply]
@GiudGiud Oh nice, didn't expect it to work just like that, I was worried about un-used data. However, in this case, I will have to generate a surface/boundary that wraps around the whole domain + use python lookup to prepare the full-list BC CSV file.

below is the test version applying random pressure BC to the LEFT boundary + drainage BC to the RIGHT
![image](https://user-images.githubusercontent.com/79562439/171789731-0f69f47b-3a9f-46ea-bfa0-8f6e866a44c1.png)

Awesome, thank you @GiudGiud!

**Comment by PEI0214** [comment]
Hello, have you found a way to add node boundary conditions?

**Reply by Traiwit** [reply]
Hi,

yes, you need to use `PropertyReadFile` > `PiecewiseConstantFromCSV` > `FunctionDirichletBC`


```

[UserObjects]
  [uobj_pwp_read_AllNodes]
    type = PropertyReadFile
    prop_file_name = 'pwp_atALLNODES_rhoGdZminus10m.csv'
    nprop = 2
    read_type = node
  []
[]

[Function]
  [fcn_pwp_initialCondition_csv]
    type = PiecewiseConstantFromCSV
    read_prop_user_object = uobj_pwp_read_AllNodes
    read_type = node
    column_number = 1
  []
[]


[BCs]
  [bc_walls]
    type = FunctionDirichletBC
    variable = porepressure_L1
    boundary = 'BOUNDSRF_WALLS'
    function = fcn_pwp_initialCondition_csv
  []
[]
```
